### PR TITLE
feat(snql) Add a function that runs purely on SnQL

### PR DESCRIPTION
**Pull request by evanh**
Add a function that converts a legacy query to SnQL and then queries the SnQL
endpoint for results (as opposed to just a dryrun). If the SnQL query fails it
will call the legacy endpoint. Hopefully once this goes into widespread use most
of the SnQL bugs will have been fixed by the dryrun code so running both queries
shouldn't happen often.

### CODE REVIEW

File: src/sentry/utils/snuba.py [hunk 1]
```diff
@@ -858,6 +865,53 @@ def _snql_query(params: Tuple[SnubaQuery, Hub, Mapping[str, str]]) -> RawResult:
         raise SnubaError(err)
 
 
+def _legacy_snql_query(params: Tuple[SnubaQuery, Hub, Mapping[str, str]]) -> RawResult:
```
Comment (wedamija) [inline src/sentry/utils/snuba.py:868]: Nit: Does `_snql_legacy_query` make more sense, since we try snql first?

**Review by wmak — APPROVED** [review]
LGTM, just one small comment about renaming variables

**Review by wedamija — APPROVED** [review]

File: src/sentry/utils/snql.py [hunk 2]
```diff
@@ -1,50 +1,80 @@
...
-    "tsdb-modelid:407",
-    "testing.test",
-}
-
-
-# Returns None (do not use SnQL), "auto" (use dataset as entity) or a specific entity to use.
-def should_use_snql(referrer: Optional[str]) -> Optional[str]:
-    if not referrer or referrer in referrer_blacklist:
+
+class SnQLOption(NamedTuple):
```
Comment (markstory) [inline src/sentry/utils/snql.py:7]: I would probably go with `SnqlOption` or `SNQLOption` as the current casing will trip someone up. :bike::house:

File: src/sentry/utils/snql.py [hunk 3]
```diff
@@ -1,50 +1,80 @@
...
-        snql_rate = options.get("snuba.snql.referrer-rate")
-        assert isinstance(snql_rate, float)
-        if random.random() <= snql_rate:
-            return entity
-
-    return None
+
+dryrun_check = ReferrerCheck(
+    blacklist=set(),
+    whitelist={
```
Comment (markstory) [inline src/sentry/utils/snql.py:47]: I would go with allowlist/denylist here.

File: src/sentry/utils/snuba.py (hunk 1 above)
Comment (evanh) [reply src/sentry/utils/snuba.py:868]: I decided to leave this, since the function takes a legacy query and outputs a snql query.

**Review by markstory — APPROVED** [review]

**Pull request merged**